measures/ResStockArgumentsPostHPXML/measure.rb

```diff
@@ -29,76 +33,215 @@ def arguments(model) # rubocop:disable Lint/UnusedMethodArgument
+    arg.setDisplayName('Load Flexibility: Random Shift (minutes)')
+    arg.setDescription('Number of minutes to randomly shift the peak period. If minutes is less than timestep, it will be assumed to be 0.')
+    arg.setDefaultValue(0)
     args << arg
 
-    arg = OpenStudio::Measure::OSArgument::makeStringArgument('output_csv_path', false)
-    arg.setDisplayName('Schedules: Output CSV Path')
-    arg.setDescription('Absolute/relative path of the csv file containing occupancy schedules. Relative paths are relative to the HPXML output path.')
+    arg = OpenStudio::Measure::OSArgument.makeStringArgument('building_id', false)
+    arg.setDisplayName('Building Unit ID')
+    arg.setDescription('The building unit number (between 1 and the number of samples).')
     args << arg
 
-    return args
+    args
   end
 
-  # define what happens when the measure is run
+  # Run the measure
   def run(model, runner, user_arguments)
     super(model, runner, user_arguments)
+    @runner = runner
+    # Use the built-in error checking
+    return false unless runner.validateUserArguments(arguments(model), user_arguments)
 
-    # use the built-in error checking
-    if !runner.validateUserArguments(arguments(model), user_arguments)
-      return false
-    end
-
-    # assign the user inputs to variables
+    # Parse user arguments and assign to variables
     args = runner.getArgumentValues(arguments(model), user_arguments)
-
-    hpxml_path = args[:hpxml_path]
-    unless (Pathname.new hpxml_path).absolute?
-      hpxml_path = File.expand_path(File.join(File.dirname(__FILE__), hpxml_path))
+    if skip_post_hpxml?(args)
+      runner.registerInfo('Skipping ResStockArgumentsPostHPXML')
+      return true
     end
-    unless File.exist?(hpxml_path) && hpxml_path.downcase.end_with?('.xml')
-      fail "'#{hpxml_path}' does not exist or is not an .xml file."
+    @args = args
+
+    @hpxml_path = args[:hpxml_path]
+    @hpxml_path = File.expand_path(@hpxml_path) unless (Pathname.new @hpxml_path).absolute?
+    raise "'#{@hpxml_path}' does not exist or is not an .xml file." unless File.exist?(@hpxml_path) && @hpxml_path.downcase.end_with?('.xml')
+
+    output_csv_path = File.dirname(@hpxml_path)
+
+    # Load HPXML
+    @hpxml = HPXML.new(hpxml_path: @hpxml_path)
+    @prng = Random.new(args[:building_id].to_i)
+    @minutes_per_step = @hpxml.header.timestep
+    max_random_shift_steps = (args[:flex_random_shift_minutes] / @minutes_per_step).to_i
+    @random_shift_steps = @prng.rand(-max_random_shift_steps..max_random_shift_steps)
+
+    # Parse the HPXML document
+    doc = XMLHelper.parse_file(@hpxml_path)
+    hpxml_doc = XMLHelper.get_element(doc, '/HPXML')
+    doc_buildings = XMLHelper.get_elements(hpxml_doc, 'Building')
+
+    # Process each building
+    doc_buildings.each_with_index do |building, index|
+      hpxml_bldg = @hpxml.buildings[index]
+      if hpxml_bldg.hvac_controls.to_a.length == 0
```

Comment:
I wish this measure was more modular where code specific to HVAC DF and EV DF could be more easily separated.